File: 백준/Gold/12912. 트리 수정/트리 수정.py
import sys
def print(*args, sep=" ", end="\n"):
    sys.stdout.write(sep.join(map(str, args)) + end)
def input():
    return sys.stdin.readline().rstrip()

def main():
    N = int(input())
    tree = [[] for _ in range(N)]
    edges = []

    for i in range(N-1):
        u, v, w = map(int, input().split())
        edges.append((u, v, w))
        tree[u].append((v, w, i))
        tree[v].append((u, w, i))

    def dfs_farthest(start, blocked_edge):
        visited = [False] * N
        stack = [(start, 0)]
        visited[start] = True

        farthest_node = start
        max_dist = 0

        while stack:
            cur, dist = stack.pop()

            if dist > max_dist:
                max_dist = dist
                farthest_node = cur
            
            for nxt, w, eid in tree[cur]:
                if eid == blocked_edge:
                    continue
                if not visited[nxt]:
                    visited[nxt] = True
                    stack.append((nxt, dist + w))
        
        return (farthest_node, max_dist)
    
    def dfs_diameter(start, blocked_edge):
        x, _ = dfs_farthest(start, blocked_edge)
        _, diameter = dfs_farthest(x, blocked_edge)

        return diameter
    
    answer = 0
    for i in range(N-1):
        u, v, w = edges[i]

        x, _ = dfs_farthest(u, i)
        _, dist_u = dfs_farthest(x, i)
        diam_u = dist_u

        y, _ = dfs_farthest(v, i)
        _, dist_v = dfs_farthest(y, i)
        diam_v = dist_v

        candidate = max(
            diam_u,
            diam_v,
            dist_u + w + dist_v
        )

        answer = max(answer, candidate)
    
    print(answer)
    
if __name__ == '__main__':
    main()

File: 백준/Gold/12912. 트리 수정/트리 수정.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
using pll = pair<ll, ll>;

int N;
vector<vector<pll>> graph;
vector<tuple<int, int, int>> edges;

pll DFS(int start, int pre, int blocked_edge){
    vector<bool> visited(N, false);
    stack<pll> st;

    st.push({start, 0});
    visited[start] = true;

    int node = start;
    ll max_dist = 0;

    while (!st.empty()){
        auto [cur, dist] = st.top(); st.pop();

        if (dist > max_dist){
            max_dist = dist;
            node = cur;
        }

        for (auto [nxt, w] : graph[cur]) {
            if (nxt == pre) continue;
            if (nxt == blocked_edge) continue;
            if (!visited[nxt]){
                visited[nxt] = true;
                st.push({nxt, dist + w});
            }
        }
    }
    return {node, max_dist};
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    cin >> N;
    graph.resize(N);

    for (int i = 0; i < N-1; i++){
        int u, v; ll w;
        cin >> u >> v >> w;
        edges.push_back({u,v,w});
        graph[u].push_back({v,w});
        graph[v].push_back({u,w});
    }

    ll answer = 0;

    for (auto [u,v,w] : edges){
        answer = max(answer, DFS(DFS(u,v,v).first, v, v).second + w + DFS(DFS(v,u,u).first, u,u).second);
    }

    cout << answer;

}
